extensions/razer/internal.m

```diff
@@ -0,0 +1,870 @@
+#import <IOKit/hidsystem/IOHIDServiceClient.h>
+
+#define USERDATA_TAG  "hs.razer"
+static LSRefTable refTable = LUA_NOREF;
+
+#define get_objectFromUserdata(objType, L, idx, tag) (objType*)*((void**)luaL_checkudata(L, idx, tag))
+
+#pragma mark - Support Functions and Classes
+
+@interface HSRazer : NSObject
+
+@property (nonatomic, strong) id    ioHIDManager;
+
+@property NSNumber*                 internalDeviceId;
+@property NSNumber*                 productId;
+
+@property int                       selfRefCount;
+@property int                       callbackRef;
+@property id                        callbackToken;
+@property int                       deviceCallbackRef;
+
+@property LSGCCanary                lsCanary;
+
+@end
+
+@implementation HSRazer
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self)
+    {
+        _callbackRef                = LUA_NOREF;
+        _deviceCallbackRef          = LUA_NOREF;
+        _callbackToken              = nil;
+        _selfRefCount               = 0;
+                
+        // Create a HID device manager
+        self.ioHIDManager = CFBridgingRelease(IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDManagerOptionNone));
+                
+        NSDictionary *matchTartarusV2   = @{ @(kIOHIDVendorIDKey): @0x1532, @(kIOHIDProductIDKey): @0x022b };
+        
+        IOHIDManagerSetDeviceMatchingMultiple((__bridge IOHIDManagerRef)self.ioHIDManager,
+                                              (__bridge CFArrayRef)@[matchTartarusV2]);
+
+        // Add our callbacks for relevant events
+        IOHIDManagerRegisterDeviceMatchingCallback((__bridge IOHIDManagerRef)self.ioHIDManager,
+                                                   HIDconnect,
+                                                   (__bridge void*)self);
+        IOHIDManagerRegisterDeviceRemovalCallback((__bridge IOHIDManagerRef)self.ioHIDManager,
+                                                  HIDdisconnect,
+                                                  (__bridge void*)self);
+        
+        IOHIDManagerRegisterInputValueCallback((__bridge IOHIDManagerRef)self.ioHIDManager, keyboardCallback, NULL);
+
+        // Start our HID manager
+        IOHIDManagerScheduleWithRunLoop((__bridge IOHIDManagerRef)self.ioHIDManager,
+                                        CFRunLoopGetMain(),
+                                        kCFRunLoopDefaultMode);
+        
+        IOHIDManagerOpen((__bridge IOHIDManagerRef)self.ioHIDManager, kIOHIDOptionsTypeNone);
+        
+        // NOTE: This gives Hammerspoon exclusive access to the Razer device, but requires `root` permissions
+        //       so it's not very helpful.
+        //IOHIDManagerOpen((__bridge IOHIDManagerRef)self.ioHIDManager, kIOHIDOptionsTypeSeizeDevice);
+        
+        CFRunLoopRun();
+    }
+    return self;
+}
+
+#pragma mark - IOKit C callbacks
+
+static void keyboardCallback(void* context, IOReturn result, void* sender, IOHIDValueRef value)
```

Comment:
@asmagill - If you get a chance, are you able to help me with some basic structural things? For example, how can I trigger the Lua callback from within this function (i.e. how can I access the `_callbackRef` from here)?

Reply:
Ignore this... worked it out, I wasn't passing the `context` correctly.